### PR TITLE
Complete probmods

### PR DESCRIPTION
**Pull request by zenna**
Probmods (probabilistic models of congition) completed by @archanarw

### CODE REVIEW

File: OmegaExamples/src/probmods/generative-models.jl [hunk 1]
```diff
@@ -0,0 +1,612 @@
...
+Programming languages are formal systems for describing what (deterministic) computation a computer should do. Modern programming languages offer a wide variety of different ways to describe computation; each makes some processes simple to describe and others more complex. However, a key tenet of computer science is that all of these languages have the same fundamental power: any computation that can be described with one programming language can be described by another. (More technically this Church-Turing thesis posits that many specific computational systems capture the set of all effectively computable procedures. These are called universal systems.)"
+
+# ╔═╡ cb28a5c4-a721-4a53-a179-adedf4f71237
+md"""
+[Omega](https://github.com/zenna/Omega.jl) is a programming language that describes probabilistic computation. 
+The key idea is that in Omega you can express stochastic processes and not just deterministic functions (like _and_).
```
Comment (robertness) [inline OmegaExamples/src/probmods/generative-models.jl:31]: - It might not be obvious to the reader that _and_ is a function.  Perhaps you might say "logical AND gate"?
- This is the first time the term "stochastic process" appears.  One should define it for the reader in a way that contrasts it with "deterministic function" and answer why it is desirable to use stochastic processes in a model.

File: OmegaExamples/src/probmods/generative-models.jl [hunk 2]
```diff
@@ -0,0 +1,612 @@
...
+# ╔═╡ 0ce18f90-f2ef-490a-a7c4-40a6ccfa214a
+md"""
+A basic object in Omega is a random variable class, from which you can generate random variables. For instance, a random variable that is of Bernoulli distribution is generated by -
+"""
+
+# ╔═╡ 354e3b5a-4d1e-43ee-b1bb-e8620fba49e7
+a = 0 ~ Bernoulli()
+
+# ╔═╡ 4c915901-e987-4552-b5f7-bc25ee6edeec
+md"Here, `Bernoulli()` is a random class and to get a member (a random variable, which can be sampled from) of the class we use `~`. So, `0 ~ Bernoulli()` is the `0`th member of Bernoulli class."
```
Comment (robertness) [inline OmegaExamples/src/probmods/generative-models.jl:43]: I assume what I'm reading here is that `a` is represents a Bernoulli random variable, not a specific Bernoulli outcome (as in 0, 1)?  If that is the case, I suggest rewording from "random class" to "a class representing a random variable"

File: OmegaExamples/src/probmods/generative-models.jl [hunk 3]
```diff
@@ -0,0 +1,612 @@
...
+The key idea is that in Omega you can express stochastic processes and not just deterministic functions (like _and_).
+"""
+
+# ╔═╡ 0ce18f90-f2ef-490a-a7c4-40a6ccfa214a
+md"""
+A basic object in Omega is a random variable class, from which you can generate random variables. For instance, a random variable that is of Bernoulli distribution is generated by -
+"""
+
+# ╔═╡ 354e3b5a-4d1e-43ee-b1bb-e8620fba49e7
+a = 0 ~ Bernoulli()
```
Comment (robertness) [inline OmegaExamples/src/probmods/generative-models.jl:40]: I think `a` here represents a random variable (object with type `Member{Bernoulli{Float64}, Int64}`).  Two comments (maybe more for @zenna)

- Perhaps a good convention would be to use capitals or camel case with the object so we remember its a random variable and not one of the RV's outcome?
- The use of "~" immediately calls into mind sampling an outcome.  Then you see the "0" and it's more confusing.  My initial though was that this is some sort of predicate for "0 is sampled from a Bernoulli."  Playing around, I see that "0" is perhaps some sort of name, because I can replace 0 with strings.  Is this 0 like `pyro.sample("0", Bernoulli())`
- Perhaps instead of "plate", this is a "causal Markov kernel", i.e. an expression representing a random variable conditioned on its direct causes?  Just a stab in the dark.
-  But then, below @archanarw says this expression represents a "plate" in a Bayes Net.  Now I'm confused all over.  When I think of plate, I think of a dimension in a model containing multiple conditionally i.i.d. RVs.  The use of 0 seems to imply an index in this dimension, but the size of the dimension itself isn't specified.  Can we contrast this "plate" idea to the plate context manager in Pyro?

I looked at this with a colleague who also has deep experience in PPLs and we're both pretty confused.  I'm not saying the semantics need changing, but perhaps they could be explained better.

File: OmegaExamples/src/probmods/generative-models.jl [hunk 4]
```diff
@@ -0,0 +1,612 @@
...
+
+# ╔═╡ 4c915901-e987-4552-b5f7-bc25ee6edeec
+md"Here, `Bernoulli()` is a random class and to get a member (a random variable, which can be sampled from) of the class we use `~`. So, `0 ~ Bernoulli()` is the `0`th member of Bernoulli class."
+
+# ╔═╡ 7d3c8448-090d-46b4-a6b6-ba20ff617be8
+md"A class is analogous to a [*plate*](https://en.wikipedia.org/wiki/Plate_notation) in Bayesian networks."
```
Comment (robertness) [inline OmegaExamples/src/probmods/generative-models.jl:46]: See comment above

File: OmegaExamples/src/probmods/generative-models.jl [hunk 5]
```diff
@@ -0,0 +1,612 @@
...
+
+1. The sample space in probability theory (often denoted $\Omega$)
+2. Space of exogenous variables in as causal graphical modelling nomenclature (often denoted $U$)
+3. The random number generator `rng::Random.AbstractRNG`.  Implementation wise, Omega.jl tries to be compatible with any model written as a function `f(rng::AbstractRNG)`
+"""
+
+# ╔═╡ d70b97b5-af55-46f9-904e-309713603155
+md"""
+Conceptually, sampling from a random variable is done in the following two distinct steps:
+1. Construct a random $ω : Ω$, where $Ω$ can be though of as the space of exogenous variables as in causal graphical modelling nomenclature or simply the random number generator, `rng::Random.AbstractRNG`.
```
Comment (robertness) [inline OmegaExamples/src/probmods/generative-models.jl:75]: - What does "Construct a random ω : Ω" mean?
- Why are you using mathtype with Greek letters?  Why no write either "Construct a random \$\omega : \Omega$\" or "Construct a random ω : Ω" without the dollar signs?  Perhaps you want the mathtype formatting?

File: OmegaExamples/src/probmods/generative-models.jl [hunk 6]
```diff
@@ -0,0 +1,612 @@
...
+# ╔═╡ 02ca238e-5a17-42dc-92a1-be975db0e7cd
+md"To understand how `randsample` works, a key concept to know is the Ω object in Omega.jl."
+
+# ╔═╡ 4ea27371-2e73-4a69-887c-b9f519fbf583
+md"""
+There are a few perspectives on what Ω is:
+
+1. The sample space in probability theory (often denoted $\Omega$)
+2. Space of exogenous variables in as causal graphical modelling nomenclature (often denoted $U$)
+3. The random number generator `rng::Random.AbstractRNG`.  Implementation wise, Omega.jl tries to be compatible with any model written as a function `f(rng::AbstractRNG)`
```
Comment (robertness) [inline OmegaExamples/src/probmods/generative-models.jl:69]: Add a link to documentation for `rng::Random.AbstractRNG`

File: OmegaExamples/src/probmods/generative-models.jl (hunk 5 above)
Comment (robertness) [reply OmegaExamples/src/probmods/generative-models.jl:75]: A note on format: why are you using mathtype with Greek letters?  Why no write either "Construct a random \$\omega : \Omega$\" or "Construct a random ω : Ω" without the dollar signs?  Perhaps you want the mathtype formatting?

File: OmegaExamples/src/probmods/generative-models.jl [hunk 7]
```diff
@@ -0,0 +1,612 @@
...
+rv = 1 ~ As
+
+# ╔═╡ 4e80ce71-6757-4908-8cd0-e382b11b70c3
+randsample(rv)
+
+# ╔═╡ 5fd22c50-4c81-41b2-a959-fe3dec00a6fd
+md"""
+##### Composite Class
+
+A class in Omega is actually just function of the form f(id, ω). Of course, you can specify your own classes simply by constructing a function.
```
Comment (robertness) [inline OmegaExamples/src/probmods/generative-models.jl:100]: I'm still hazy on what "id"'s are

File: OmegaExamples/src/probmods/generative-models.jl [hunk 8]
```diff
@@ -0,0 +1,612 @@
...
+"""
+
+# ╔═╡ 8ac4cad4-eb4c-4e33-815c-e365668b3810
+μ = 2 ~ StdNormal{Float64}()
+
+# ╔═╡ 5fcf6c97-6f81-4be2-b27a-25b2abcabca1
+x(id, ω) = (id ~ Normal(μ(ω), 1))(ω)
```
Comment (robertness) [inline OmegaExamples/src/probmods/generative-models.jl:107]: Ah, so μ, an RV, is callable with ω as an argument.  This is the first time I'm seeing this pattern in this tutorial, as is the composite non-standard Normal you construct.  Is it true that all primitive classes don't take ω as an argument and all composite classes do?  If so, say so.

The use of ω as an exposed thread that ties together a sequence of expressions is a fundamentally unique element Omega.  I suggest at least explicitly highlighting that fact here.  Explaining it here would be ideal, but perhaps it is explained elsewhere in the docs.  If so, please add a link to that doc.

File: OmegaExamples/src/probmods/generative-models.jl [hunk 9]
```diff
@@ -0,0 +1,612 @@
...
+
+# ╔═╡ 1b4d9047-abb0-4fb7-ab72-0a4f790b1515
+x_ = 3 ~ x
+
+# ╔═╡ 0fe48b01-e908-4e6c-8c4c-b68cf8c172dc
+md"`x_` is a random variable of the class `x`"
```
Comment (robertness) [inline OmegaExamples/src/probmods/generative-models.jl:113]: At this point, as an informed reader, I'm seeing that μ and x have the same type and x_ is a function (and I think all are callable?) .  I'm really wishing there were at least some naming conventions to help me distinguish these concepts in my head.

File: OmegaExamples/src/probmods/generative-models.jl [hunk 10]
```diff
@@ -0,0 +1,612 @@
...
+
+# ╔═╡ 0a2c3c83-8a67-4666-80c0-f580233dc019
+randsample((μ, x_))
+
+# ╔═╡ f2a20f84-b2e6-4e6c-ab51-6e7ea045954e
+md"Alternatively, `@joint` is used to create the joint distribution of variables:"
```
Comment (robertness) [inline OmegaExamples/src/probmods/generative-models.jl:119]: I think `@joint` is a Julia macro.  I think it would be good to say so, as in "Alternatively, the macro `@joint` is used to ...", and to link to the code for `@joint` in the docs.

File: OmegaExamples/src/probmods/generative-models.jl [hunk 11]
```diff
@@ -0,0 +1,612 @@
...
+
+# ╔═╡ 7d5587d1-c4b4-48f0-971c-677fbe268e73
+joint = @joint μ x_
+
+# ╔═╡ ae0ad1df-ab2a-4704-a121-623e05f079da
+randsample(joint)
+
+# ╔═╡ 0ae343fa-bc62-48ac-a34e-d21028531530
+md"""
+Every time we create a new (independent) random variable, we need to provide a unique id to it. We could instead use `@~` to generate `id`s automatically. For example, to simulate a (possibly biased) coin toss, we can simply write:
```
Comment (robertness) [inline OmegaExamples/src/probmods/generative-models.jl:129]: Again, it would be good to explain the id thing and the ω thing, why they exist and what they do, as early in the doc as possible, or link to a explainer as early as possible, since they are going to be a source of confusion for newcomers to Omega.

File: OmegaExamples/src/probmods/generative-models.jl [hunk 12]
```diff
@@ -0,0 +1,612 @@
...
+
+# ╔═╡ 9504162b-52e1-476d-8218-0afc2555e278
+viz(randsample(coin_toss, 1000))
+
+# ╔═╡ 15920581-2258-42e1-ba73-da82b2baa7e0
+md"As you can see, the result is an approximately uniform distribution over true(1) and false(0).
+This way we can construct more complex expressions that describe more complicated random variables. For instance, here we describe a process that samples a number adding up several independent Bernoulli distributions:"
+
+# ╔═╡ 3f04b226-a38a-4804-8691-59332bfc8728
+b_sum = (@~ Bernoulli()) +ₚ (@~ Bernoulli()) +ₚ (@~ Bernoulli())
```
Comment (robertness) [inline OmegaExamples/src/probmods/generative-models.jl:149]: +ₚ is not defined for me:
 
image

I'm using Julia 1.7.0 and I followed the instructions @archanarw gave me.
image

Please let me know how to fix it and then I'll continue with the PR.

**Review by robertness — CHANGES_REQUESTED** [review]
I've started to go through this document with a colleague. Both of us are familiar with ppls and generative models.  I'm making suggestions for how to make the document easier to understand for us -- knowledgeable people who are new to Omega.

I think that should be the target audience for this doc -- people who already know why one would use a generative model to solve a problem, and who know what R.V.s and distributions and Bayes rule mean, but who are new to Omega.

@archanarw this PR is way too big.  Let's break this up into a series of small PRs.  Use this complete-probmods branch as a place to store all the work you've done, but create new branches to merge.  Specifically, let's do the following steps:

1. Close this PR without merging
2. Create a new branch called generative-models (or something like that) **from master**.
3. From the complete-probmods branch ONLY copy probmods/generative-models.jl and ONLY the files necessary for generative-models.jl to run.
4. Make sure the whole document runs without error.
5. Create a PR for the new generative-models branch.  Start working on my edits in that PR. 
6. After I changes, merge this doc to master.  When we merge this to master, it will be like [the first chapter of probmods](https://probmods.org/chapters/generative-models.html) is done.
8. Delete the new generative models branch after merging.
9. Pull from master and make sure generative-models.jl still runs without error.
10. Create a new branch for the next chapter.  We'll continue this chapter-by-chapter.

File: OmegaExamples/src/probmods/generative-models.jl (hunk 5 above)
Comment (archanarw) [reply OmegaExamples/src/probmods/generative-models.jl:75]: I'm not sure it's necessary for the user in the first place since it describes internals of Omega, it's not something the user needs to do, `randsample` does it. Should I remove this portion? 
I included it because we write functions of `\omega::\Omega`  to construct random variables (which `randsample` can be used on) and I wasn't sure how to introduce that.

File: OmegaExamples/src/probmods/generative-models.jl (hunk 9 above)
Comment (archanarw) [reply OmegaExamples/src/probmods/generative-models.jl:113]: μ and x_ are of the same type and we can use `randsample` to sample from them. x can not be sampled from since it's a class (so an instance, @~ x, can be used with `randsample`). 

μ and x_ are functions of only one argument (ω) and x is a function of id and ω. So any function of ω is a random variable in Omega and functions of the form f(id, ω) are random classes.
